indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java

```diff
@@ -1454,20 +1454,23 @@ public void runInternal()
 
       checkCurrentTaskState();
 
-      synchronized (stateChangeLock) {
-        // if supervisor is not suspended, ensure required tasks are running
-        // if suspended, ensure tasks have been requested to gracefully stop
-        if (stateManager.getSupervisorState().getBasicState().equals(SupervisorStateManager.BasicState.STOPPING)) {
-          // if we're already terminating, don't do anything here, the terminate already handles shutdown
-          log.info("[%s] supervisor is already stopping.", dataSource);
-        } else if (!spec.isSuspended()) {
-          log.info("[%s] supervisor is running.", dataSource);
-
-          stateManager.maybeSetState(SeekableStreamSupervisorStateManager.SeekableStreamState.CREATING_TASKS);
-          createNewTasks();
-        } else {
-          log.info("[%s] supervisor is suspended.", dataSource);
-          gracefulShutdownInternal();
+      // If supervisor is already stopping, don't contend for stateChangeLock since the block can be skipped
+      if (!stateManager.getSupervisorState().getBasicState().equals(SupervisorStateManager.BasicState.STOPPING)) {
```

Comment:
Is there a race here? Could the state be set to STOPPING between this check and the entry into `synchronized (stateChangeLock)`?

Reply:
There is an additional check within the synchronized block to see if it has been set to STOPPING.
The first check was added with the intent to prevent the need to enter that block once the supervisor has begun stopping.
One such scenario is when a supervisor tries to terminate just before the CREATING_TASKS phase which can be reproduced by terminating a supervisor, after a period of its configured startDelay, right after creation.

Reply:
Could you please share a better alternative to avoid waiting to enter the synchronized block after the supervisor has begun?
The lock leads to a delay in the run itself and gracefulShutdown times out frequently in the specified case and there isn't anything to do in the block in such cases